### DISCUSSION TITLE
Unit test file uplload

### DISCUSSION THREAD

**Discussion by PeterAshfordPEL** [mirumee/ariadne#729] [Q&A] [answered]
Hi there! 
I'm unit testing ariadne graphql mutations using calls to graphql_sync which is working really well. However I have one mutation that takes a file upload. I can test this from the command line using curl, but I'd like to be able to add a python unit test for the upload to our standard test suite.
My problem is that I don't know how to write a test that exercises the file upload. I've been Googling for answers with no success.

**Comment by rafalp** [comment, marked as answer]
Here's test that does this for plain WSGI server, maybe this will help you somehow?

https://github.com/mirumee/ariadne/blob/master/tests/wsgi/test_query_execution.py#L170

**Reply by PeterAshfordPEL** [reply]
Thanks, that helps a lot!

**Comment by fdemian** [comment]
Hi everyone. Sorry for reviving this discussion, but I have a similar problem testing an application with starlette's test client. Here's what I'm trying to do 

 ```
     # Upload profile pic
      update_query = """
        mutation UploadAvatar($image: Upload!, $id: Int!) {
           uploadUserImage(image: $image, id: $id) {
             id
             ok
           }
      }
      """
      files_map = {"0": ["variables.image"]}
      upload_data = {
         "operationName": "UploadAvatar",
         "query": update_query.replace('\n', '').strip(),
         "variables": {
           "image": None,
           "id": 1
         },
         "map": files_map
       }

       file_cont = get_sample_file_contents()
       files = {"file": (file_cont['name'], file_cont['content'], "image/png")}
       headers = {
         'content-type': 'application/json',
         'access_token': access_token
       }
       upload_resp = client.post("/api/graphql", data=json.dumps(upload_data), headers=headers, files=files)
```

I'm getting the following error

`Variable '$image' of non-null type 'Upload!' must not be null.`

 It seems like I can't starlette's test client to correctly send the files mapping.  Is there a way to do this with starlette's test client?
